tests/tools/app_store_connect/test_publish_action.py

```diff
@@ -65,6 +65,107 @@ def test_publish_action_with_username_and_password(_mock_altool, namespace_kwarg
         mock_get_packages.assert_called_with(patterns)
 
 
+def test_publish_action_testflight(namespace_kwargs):
+    namespace_kwargs.update({'action': 'publish'})
+
+    cli_args = argparse.Namespace(**namespace_kwargs)
+    with mock.patch.object(AppStoreConnect, 'find_paths') as mock_find_paths, \
+            mock.patch.object(AppStoreConnect, '_get_publishing_application_packages') as mock_get_packages:
+        mock_get_packages.return_value = []
+        mock_find_paths.return_value = []
+
+        patterns = [pathlib.Path('path.pattern')]
+        AppStoreConnect.from_cli_args(cli_args).publish(
+            application_package_path_patterns=patterns,
+            apple_id='name@example.com',
+            app_specific_password=Types.AppSpecificPassword('xxxx-yyyy-zzzz-wwww'),
+            submit_to_testflight=True,
+        )
+        mock_get_packages.assert_called_with(patterns)
+
+
+@pytest.mark.parametrize('localization_arguments', [{'whats_new': "What's new"}, {'locale': 'en-GB'}])
+def test_publish_action_testflight_with_localization_with_missing_key(localization_arguments, namespace_kwargs):
+    namespace_kwargs.update({'action': 'publish'})
+    cli_args = argparse.Namespace(**namespace_kwargs)
+    patterns = [pathlib.Path('path.pattern')]
+
+    with pytest.raises(IOError) as error_info:
+        AppStoreConnect.from_cli_args(cli_args).publish(
+            application_package_path_patterns=patterns,
+            apple_id='name@example.com',
+            app_specific_password=Types.AppSpecificPassword('xxxx-yyyy-zzzz-wwww'),
+            submit_to_testflight=True,
+            **localization_arguments,
+        )
+
+    assert str(error_info.value) \
+        == 'Both --locale and --whats-new need to be defined for providing beta build localization'
+
+
+def test_publish_action_with_localization_and_no_testflight_submission(namespace_kwargs):
+    namespace_kwargs.update({'action': 'publish'})
+    cli_args = argparse.Namespace(**namespace_kwargs)
+    patterns = [pathlib.Path('path.pattern')]
+
+    with pytest.raises(IOError) as error_info:
+        AppStoreConnect.from_cli_args(cli_args).publish(
+            application_package_path_patterns=patterns,
+            apple_id='name@example.com',
+            app_specific_password=Types.AppSpecificPassword('xxxx-yyyy-zzzz-wwww'),
+            locale='en-GB',
+            whats_new="What's new",
+        )
+
+    assert str(error_info.value) \
+        == '--testflight flag is required for providing beta build localization'
+
+
+def test_publish_action_testflight_with_localization(namespace_kwargs):
+    class MockBuild:
+        def __init__(self, build_id):
+            self.id = build_id
+
+    class MockIpa:
```

Comment:
It is used once elsewhere in this file. Why not extract it to top level definition to avoid code duplication?